File: sinkworld/tentacle/csharp/Surface.cs
using System;
using System.Drawing;
namespace SinkWorld {

public class Surface {
	public static Literal fontDefault = null;

	internal Graphics g;
	internal FontDescription font;
	internal Font fontHandle;
	internal int ascent = 1;
	internal SolidBrush textBrush = new SolidBrush(Color.Black);
	internal Pen linePen = new Pen(Color.Black);

	internal RGBColor currentColor = new RGBColor(0x1000000);
	internal int x = 0;
	internal int y = 0;
	internal StringFormat sf = null;

	internal Surface(Graphics g_) {
		g = g_;
		sf = new StringFormat(StringFormat.GenericTypographic);
		sf.FormatFlags = StringFormatFlags.MeasureTrailingSpaces;
		//g.TextRenderingHint = System.Drawing.Text.TextRenderingHint.AntiAlias;
		//g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.AntiAlias;
		g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.None;
	}

	public virtual void SetFont(FontDescription font_) {
		if (font != font_) {
			if ((font == null) || (!font.Equals(font_))) {
				font = font_;
				System.String fontName = "Verdana";
				try {
					int maxSize = font.font.Length() * 3;
					byte[] nameBytes = new byte[maxSize];
					int lenUTF8 = UniConversion.UTF8FromUTF32(
						nameBytes, nameBytes.Length,
						font.font.Data(), font.font.Length());
					//~ byte[] nameBytes = font.font.Value();
					fontName = System.Text.Encoding.UTF8.GetString(nameBytes, 0, lenUTF8);
				}
				catch (System.IO.IOException) {
					System.Console.Out.WriteLine("Failed to convert");
				}
				catch (System.NullReferenceException) {
					System.Console.Out.WriteLine("Null to convert");
				}
				//~ System.out.println("SetFont " + fontName + " " + font.height);
				int fs = 0;
				if (font.bold)
					fs |= (int) FontStyle.Bold;
				if (font.italics)
					fs |= (int) FontStyle.Italic;
				fontHandle = new Font(fontName, font.height, (FontStyle) fs);
				ascent = TextAscent();
			}
		}
	}
	public virtual void SetTextColor(RGBColor color) {
		SetPen(color);
	}
	public virtual void SetPen(RGBColor color) {
		if (currentColor.NotEqual(color)) {
			Color col = Color.FromArgb((int)color.GetRed(), (int)color.GetGreen(), (int)color.GetBlue());
			textBrush = new SolidBrush(col);
			linePen = new Pen(col, 1.0f);
			currentColor.co = color.co;
		}
	}
	public virtual void Free() {
	}
	public virtual void FillRectangle(SRectangle rc, RGBColor color) {
		SetPen(color);
		g.FillRectangle(textBrush, rc.left, rc.top, rc.right - rc.left, rc.bottom - rc.top);
	}
	public virtual void FrameRectangle(SRectangle rc, RGBColor color) {
		SetPen(color);
		g.DrawRectangle(linePen, rc.left, rc.top, rc.right - rc.left, rc.bottom - rc.top);
	}
	public virtual void Polygon(Point []pts, int npts, RGBColor fore, int alphaFill, int alphaOutline) {
		// Convert from SinkWorld Point array to system Point array
		System.Drawing.Point[] poly = new System.Drawing.Point[npts];
		for (int i=0; i<npts; i++) {
			poly[i] = new System.Drawing.Point(pts[i].x, pts[i].y);
		}
		Color colFill = Color.FromArgb(alphaFill, (int)fore.GetRed(), (int)fore.GetGreen(), (int)fore.GetBlue());
		g.FillPolygon(new SolidBrush(colFill), poly);
		Color colOutline = Color.FromArgb(alphaOutline, (int)fore.GetRed(), (int)fore.GetGreen(), (int)fore.GetBlue());
		g.DrawPolygon(new Pen(colOutline, 0.0f), poly);
	}
	public virtual void MoveTo(int x_, int y_) {
		x = x_;
		y = y_;
	}
	public virtual void LineTo(int x_, int y_) {
		g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.AntiAlias;
		g.DrawLine(linePen, x, y, x_, y_);
		g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.None;
		x = x_;
		y = y_;
	}
	public virtual void SetTransparent(bool transparent) {
	}
	public virtual void xMeasureWidths(byte[] s, int len, TextPositions positions, int enc) {
		try {
			String sg = System.Text.Encoding.UTF8.GetString(s, 0, len);
			int sLength = sg.Length;
			SizeF sz = g.MeasureString(sg, fontHandle, 2000, sf);
			float maxWidth = sz.Width;
			float x = 0;
			float []xPositions = new float[sLength];
			for (int i = 0; i < sLength; i++) {
				char ch = sg[i];
				sz = g.MeasureString(new String(ch, 1), fontHandle, 2000, sf);
				x += sz.Width;
				xPositions[i] = x;
			}
			float ratio = maxWidth / x;
			for (int i = 0; i < sLength; i++) {
				char ch = sg[i];
				int xPixel = (int)(xPositions[i] * ratio);
				positions.Add(xPixel);
				if (ch >= 0x80)
					positions.Add(xPixel);
				if (ch >= 0x800)
					positions.Add(xPixel);
			}
		}
		catch (System.IO.IOException) {
			System.Console.Out.WriteLine("Failed to convert");
		}
	}
	public virtual void MeasureWidths(byte[] s, int len, TextPositions positions, int enc) {
		// A bug in GDI+ means that the string has to be broken up into blocks
		// of 32 or less.
		int maxMeasurePerCall = 32;
		try {
			String sAll = System.Text.Encoding.UTF8.GetString(s, 0, len);
			int sAllLength = sAll.Length;
			for (int start=0; start<sAllLength; start+=maxMeasurePerCall) {
				int sLength = sAllLength - start;
				if (sLength > maxMeasurePerCall) {
					sLength = maxMeasurePerCall;
				}
				CharacterRange []characterRanges = new CharacterRange[sLength];
				for (int i = 0; i<sLength; i++) {
					// Another problem with measuring trailing spaces means
					// that this doesn't work with a range containing a single space.
					if (sAll[start + i] == ' ')
						characterRanges[i] = new CharacterRange(start, i+1);
					else
						characterRanges[i] = new CharacterRange(start+i, 1);
				}
				sf.SetMeasurableCharacterRanges(characterRanges);
				RectangleF layoutRect = new RectangleF(0.0f, 0.0f, float.MaxValue, float.MaxValue);
				Region[] regions = new Region[sLength];
				regions = g.MeasureCharacterRanges(sAll, fontHandle, layoutRect, sf);

				for (int i = 0; i < sLength; i++) {
					char ch = sAll[start+i];
					float xOfCharRight = (regions[i].GetBounds(g).Right);
					int xPixel = (int)(xOfCharRight);
					positions.Add(xPixel);
					if (ch >= 0x80)
						positions.Add(xPixel);
					if (ch >= 0x800)
						positions.Add(xPixel);
				}
			}
		}
		catch (System.IO.IOException) {
			System.Console.Out.WriteLine("Failed to convert");
		}
	}
	public virtual void MeasureWidths(SplitText st, int position, int len, TextPositions positions) {
		byte[] s = new byte[len * 3];
		int lenUTF8 = st.RetrieveUTF8(position, s, len);
		MeasureWidths(s, lenUTF8, positions, st.GetEncoding());
	}
	public virtual void PaintText(byte[] s, int len, SRectangle rcw, int ascentLine, int enc) {
		try {
			System.String sg = System.Text.Encoding.UTF8.GetString(s, 0, len);
			g.DrawString(sg, fontHandle, textBrush, rcw.left, rcw.top + ascentLine - ascent, sf);
		}
		catch (System.IO.IOException) {
			System.Console.Out.WriteLine("Failed to convert");
		}
	}
	public virtual void PaintText(SplitText st, int position, int len, SRectangle rcw, int ascentLine) {
		try {
			byte[] s = new byte[len * 3];
			int lenUTF8 = st.RetrieveUTF8(position, s, len);
			System.String sg = System.Text.Encoding.UTF8.GetString(s, 0, lenUTF8);
			g.DrawString(sg, fontHandle, textBrush, rcw.left, rcw.top + ascentLine - ascent, sf);
		}
		catch (System.IO.IOException) {
			System.Console.Out.WriteLine("Failed to convert");
		}
	}
	public virtual int TextAscent() {
		FontFamily fontFamily = fontHandle.FontFamily;
		int ascentCell = fontFamily.GetCellAscent(fontHandle.Style);
		int ascentPixel = (int)fontHandle.Size * ascentCell / fontFamily.GetEmHeight(fontHandle.Style);
		return ascentPixel;
	}
	public virtual int TextDescent() {
		FontFamily fontFamily = fontHandle.FontFamily;
		int ascent = fontFamily.GetCellAscent(fontHandle.Style);
		int ascentPixel = (int)fontHandle.Size * ascent / fontFamily.GetEmHeight(fontHandle.Style);
		return (int)fontHandle.GetHeight() - ascentPixel;
	}
	public virtual void SetClip(SRectangle rc) {
		Rectangle clipRect = new Rectangle(
			rc.left, rc.top, rc.right - rc.left, rc.bottom - rc.top);
		Region clipRegion = new Region(clipRect);
		g.Clip = clipRegion;
	}
	static Surface() {
		fontDefault = new Literal("Verdana");
	}
	public const int lineExtra = 0;
}

}

File: sinkworld/tentacle/java/Surface.java
package SinkWorld;

import java.awt.*;

class FontCacheEntry {
	public FontDescription font;
	public Font fontHandle;
}

class FontCache {
	public int size;
	public FontCacheEntry []entries;

	FontCache() {
		size = 0x20;
		entries = new FontCacheEntry[size];
	}
	private int Hash(FontDescription font) {
		return font.Hash() & 0x1F;
	}
	Font GetFont(FontDescription font) {
		FontCacheEntry fce = entries[Hash(font)];
		if (fce == null)
			return null;
		else if (fce.font.Equals(font))
			return fce.fontHandle;
		else
			return null;
	}
	void SetFont(FontDescription font, Font fontHandle) {
		int probe = Hash(font);
		entries[probe] = new FontCacheEntry();
		entries[probe].font = font;
		entries[probe].fontHandle = fontHandle;
	}
}

public class Surface {

	public static Literal fontDefault=null;
	static {
		fontDefault = new Literal("Verdana");
	}

	static int totals=0;
	static int misses = 0;

	Graphics g;
	FontDescription font;
	Font fontHandle;
	FontCache fc = new FontCache();
	FontMetrics fm;
	RGBColor currentColor = new RGBColor(0x1000000);
	int x = 0;
	int y = 0;

	Surface(Graphics g_) {
		g = g_;
	}

	public void SetFont(FontDescription font_) {
		if (font != font_) {
			if ((font == null) || (!font.Equals(font_))) {
				// Only create a new font when different from cached
				font = font_;
				fontHandle = fc.GetFont(font);
				totals++;
				if (fontHandle == null) {
					misses++;
					//~ if (misses % 500 == 0) {
						//~ System.err.println("Font Cache total " + totals +
							//~ "  misses " + misses + "  proportion " +
							//~ ((float)(misses) / totals));
					//~ }
					// Not in cache so create
					String stringFont = "Verdana";
					try {
						int maxSize = font.font.Length() * 3;
						byte nameBytes[] = new byte[maxSize];
						int lenUTF8 = UniConversion.UTF8FromUTF32(
							nameBytes, nameBytes.length,
							font.font.Data(), font.font.Length());
						stringFont = new String(nameBytes, 0, lenUTF8, "UTF8");
					} catch (java.io.UnsupportedEncodingException e) {
						System.out.println("Failed to convert");
					} catch (java.lang.NullPointerException e) {
						System.out.println("Null to convert");
					}
					int fs=0;
					if (font.bold)
						fs |= Font.BOLD;
					if (font.italics)
						fs |= Font.ITALIC;
					fontHandle = new Font(stringFont, fs, font.height+2);
					fc.SetFont(font, fontHandle);
				}
				g.setFont(fontHandle);
				fm = null;
			}
		}
	}
	public void SetTextColor(RGBColor color) {
		SetPen(color);
	}
	public void SetPen(RGBColor color) {
		if (currentColor.NotEqual(color)) {
			g.setColor(new java.awt.Color((int)color.GetRed(),
				(int)color.GetGreen(), (int)color.GetBlue()));
			currentColor.co = color.co;
		}
	}
	public void Free() {
	}
	public void FillRectangle(SRectangle rc, RGBColor color) {
		SetPen(color);
		g.fillRect(rc.left, rc.top, rc.right - rc.left, rc.bottom - rc.top);
	}
	public void FrameRectangle(SRectangle rc, RGBColor color) {
		SetPen(color);
		g.drawRect(rc.left, rc.top, rc.right - rc.left - 1, rc.bottom - rc.top - 1);
	}
	public void Polygon(Point []pts, int npts, RGBColor fore, int alphaFill, int alphaOutline) {
		// Convert from SinkWorld Point array to parallel int arrays
		Polygon poly = new Polygon();
		for (int i=0; i<npts; i++) {
			poly.addPoint(pts[i].x, pts[i].y);
		}
		g.setColor(new java.awt.Color((int)fore.GetRed(), (int)fore.GetGreen(), (int)fore.GetBlue(), alphaFill));
		g.fillPolygon(poly);
		g.setColor(new java.awt.Color((int)fore.GetRed(), (int)fore.GetGreen(), (int)fore.GetBlue(), alphaOutline));
		g.drawPolygon(poly);
		
		// Ensure synchronized
		g.setColor(new java.awt.Color((int)fore.GetRed(),
			(int)fore.GetGreen(), (int)fore.GetBlue()));
		currentColor.co = fore.co;
	}
	public void MoveTo(int x_, int y_) {
		x = x_;
		y = y_;
	}
	public void LineTo(int x_, int y_) {
		g.drawLine(x, y, x_, y_);
		x = x_;
		y = y_;
	}
	public void SetTransparent(boolean transparent) {
	}
	public void MeasureWidths(byte []s, int len, TextPositions positions, int enc) {
		try {
			String sg = new String(s, 0, len, "UTF8");
			if (fm == null)
				fm = g.getFontMetrics(fontHandle);
			int x = 0;
			for (int i=0;i<sg.length(); i++) {
				char ch = sg.charAt(i);
				//~ if (ch > '\177') {
					//~ System.out.println("Measuring " + Character.getNumericValue(ch));
				//~ }
				x += fm.charWidth(ch);
				positions.Add(x);
				if (ch >= 0x80)
					positions.Add(x);
				if (ch >= 0x800)
					positions.Add(x);
			}
		} catch (java.io.UnsupportedEncodingException e) {
			System.out.println("Failed to convert");
		}
	}
	public void MeasureWidths(SplitText st, int position, int len, TextPositions positions) {
		byte s[] = new byte[len * 3];
		int lenUTF8 = st.RetrieveUTF8(position, s, len);
		MeasureWidths(s, lenUTF8, positions, st.GetEncoding());
	}
	public void PaintText(byte []s, int len, SRectangle rcw, int ascent, int enc) {
		try {
			String sg = new String(s, 0, len, "UTF8");
			g.drawString(sg, rcw.left, rcw.top+ascent);
		} catch (java.io.UnsupportedEncodingException e) {
			System.out.println("Failed to convert");
		}
	}
	public void PaintText(SplitText st, int position, int len, SRectangle rcw, int ascent) {
		try {
			byte s[] = new byte[len * 3];
			int lenUTF8 = st.RetrieveUTF8(position, s, len);
			String sg = new String(s, 0, lenUTF8, "UTF8");
			g.drawString(sg, rcw.left, rcw.top+ascent);
		} catch (java.io.UnsupportedEncodingException e) {
			System.out.println("Failed to convert");
		}
	}
	public int TextAscent() {
		if (fm == null)
			fm = g.getFontMetrics(fontHandle);
		return fm.getAscent();
	}
	public int TextDescent() {
		if (fm == null)
			fm = g.getFontMetrics(fontHandle);
		return fm.getDescent();
		//~ return fm.getLeading() + fm.getDescent();
	}
	void SetClip(SRectangle rc) {
		g.setClip(rc.left, rc.top, rc.right - rc.left, rc.bottom - rc.top);
	}
	public static final int lineExtra = 0;
};

